### PR TITLE
PWX-4411: Convert node IP to ID if necessary for API calls.

### PR DESCRIPTION
**Pull request by jrivera-px**
**What this PR does / why we need it**:
PWX-4411:  API calls which take in node id should be also be able to handle the node ip also.  Check input to see if its an ip and convert it to the corresponding node ID.    This was a customer request and allows for a more flexible API. 

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:
Add an IP to uuid lookup to :
restoreCloudSnap,
inspectNode,
deleteNode,
createVolume,
updateVolumeReplication

### CODE REVIEW

File: api/server/volume.go [hunk 1]
```diff
@@ -66,6 +67,47 @@ func (vd *volAPI) parseID(r *http.Request) (string, error) {
...
+				return nodeIds, err
+			}
+			nodeIds = append(nodeIds, id)
+		}
+	}
+
+	return nodeIds, err
+}
+
+func (vd *volAPI) replicaSpecIPtoIds(rspecRef *api.ReplicaSet) error {
```
Comment (lpabon) [inline api/server/volume.go]: Add a comment and change function name to updateXXX

File: api/server/cluster.go [hunk 2]
```diff
@@ -171,6 +171,47 @@ func (c *clusterApi) gossipState(w http.ResponseWriter, r *http.Request) {
...
+//   in: path
+//   description: cluster node ip or id
+//   required: true
+//   type: string
+// responses:
+//   '200':
+//      description: cluster node ID
+//      schema:
+//         type: string
+func (c *clusterApi) getnodeidfromip(w http.ResponseWriter, r *http.Request) {
```
Comment (sangleganesh) [inline api/server/cluster.go]: nit: camel case ?

File: api/server/volume.go [hunk 3]
```diff
@@ -66,6 +67,49 @@ func (vd *volAPI) parseID(r *http.Request) (string, error) {
 	return "", fmt.Errorf("could not parse snap ID")
 }
 
+func (vd *volAPI) nodeIPtoIds(nodes []string) ([]string, error) {
+	nodeIds := make([]string, 0)
```
Comment (sangleganesh) [inline api/server/volume.go]: nit: nodeIds := make([]string, len(nodes))

File: cluster/manager.go [hunk 4]
```diff
@@ -1398,6 +1424,13 @@ func (c *ClusterManager) Remove(nodes []api.Node, forceRemove bool) error {
 	inQuorum := !(c.selfNode.Status == api.Status_STATUS_NOT_IN_QUORUM)
 
 	for _, n := range nodes {
+
+		if id, cerr := c.GetNodeIdFromIp(n.Id); cerr == nil {
```
Comment (sangleganesh) [inline cluster/manager.go]: i think "n" is a copy of api.Node, prefer to change nodes[i] if that is referred elsewhere.

File: api/server/cluster.go [hunk 5]
```diff
@@ -171,6 +171,47 @@ func (c *clusterApi) gossipState(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		c.sendError(c.name, method, w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	vars := mux.Vars(r)
+	nodeIP, ok := vars["idip"]
+
+	if !ok || nodeIP == "" {
+		c.sendError(c.name, method, w, "Missing id param", http.StatusBadRequest)
```
Comment (adityadani) [inline api/server/cluster.go]: nit: Missing input param. 
The input can be id as well as ip right?

File: cluster/manager.go [hunk 6]
```diff
@@ -266,6 +268,30 @@ func (c *ClusterManager) GetData() (map[string]*api.Node, error) {
 	return nodes, nil
 }
 
+func (c *ClusterManager) nodeIdFromIp(idIp string) (string, error) {
+	// Caller's responsibility to lock the access to the NodeCache.
+	for _, n := range c.nodeCache {
+		if n.DataIp == idIp || n.MgmtIp == idIp {
+			logrus.Infof("Node IP: " + idIp + " maps to ID: " + n.Id)
```
Comment (adityadani) [inline cluster/manager.go]: nit: may be you want to remove his log line as it will log for every http request which makes call to this API.

**Pull request merged**